src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs

```diff
@@ -0,0 +1,197 @@
+
+namespace Tester.TestStreamProviders.Generator
+{
+    public class GeneratorPooledCache : IQueueCache
+    {
+        private readonly IObjectPool<FixedSizeBuffer> bufferPool;
+        private readonly ICacheDataAdapter<GeneratedBatchContainer, CachedMessage> operations;
+        private readonly PooledQueueCache<GeneratedBatchContainer, CachedMessage> cache;
+
+        public GeneratorPooledCache(IObjectPool<FixedSizeBuffer> bufferPool)
+        {
+            if (bufferPool == null)
+            {
+                throw new ArgumentNullException("bufferPool");
+            }
+            this.bufferPool = bufferPool;
+            operations = new CacheDataAdapter(bufferPool, disposable => cache.Purge(disposable));
+            cache = new PooledQueueCache<GeneratedBatchContainer, CachedMessage>(operations);
+        }
+
+        // For fast GC this struct should contain only value types.  I includes streamNamespace because I'm lasy and this is test code, but it should not be in here.
+        private struct CachedMessage
+        {
+            public Guid StreamGuid;
+            public string StreamNamespace;
+            public long SequenceNumber;
+            public ArraySegment<byte> Payload;
+        }
+
+        private class CacheDataAdapter : ICacheDataAdapter<GeneratedBatchContainer, CachedMessage>
+        {
+            private readonly IObjectPool<FixedSizeBuffer> bufferPool;
+            private readonly Action<IDisposable> purgeAction;
+            private FixedSizeBuffer currentBuffer;
+
+            public CacheDataAdapter(IObjectPool<FixedSizeBuffer> bufferPool, Action<IDisposable> purgeAction)
+            {
+                if (bufferPool == null)
+                {
+                    throw new ArgumentNullException("bufferPool");
+                }
+                if (purgeAction == null)
+                {
+                    throw new ArgumentNullException("purgeAction");
+                }
+                this.bufferPool = bufferPool;
+                this.purgeAction = purgeAction;
+            }
+
+            public void QueueMessageToCachedMessage(ref CachedMessage cachedMessage, GeneratedBatchContainer queueMessage)
+            {
+                cachedMessage.StreamGuid = queueMessage.StreamGuid;
+                cachedMessage.StreamNamespace = queueMessage.StreamNamespace;
+                cachedMessage.SequenceNumber = queueMessage.RealToken.SequenceNumber;
+                cachedMessage.Payload = SerializeMessageIntoPooledSegment(queueMessage);
+            }
+
+            // Placed object message payload into a segment from a buffer pool.  When this get's too big, older blocks will be purged
+            private ArraySegment<byte> SerializeMessageIntoPooledSegment(GeneratedBatchContainer queueMessage)
+            {
+                // serialize payload
+                byte[] serializedPayload = SerializationManager.SerializeToByteArray(queueMessage.Payload);
```

Comment:
This is dependent on queue data type.  If the queue data payload is available as raw binary data like in Kafka or EventHub this is not necessary, but in this case, the generated generates fully populated objects, so we need to serialize them.  In any case, this cost is minimal.  Gen0 allocations are, unless excessive, insignificant.  Given all the other overhead, I doubt optimizing this would be noticeable.